packages/runtime-common/search-index.ts

```diff
@@ -41,36 +146,91 @@ export class SearchIndex {
       path: LocalPath,
       opts?: { withFallbacks?: true }
     ) => Promise<{ content: string; lastModified: number } | undefined>,
-    getVisitor: (
-      _fetch: typeof fetch,
-      staticResponses: Map<string, string>
-    ) => (url: string) => Promise<string>
+    runner: IndexRunner
   ) {
-    this.#currentRun = new CurrentRun({
-      realm,
-      reader: { readdir, readFileAsText },
-      getVisitor,
-    });
-  }
-
-  async run() {
-    await CurrentRun.fromScratch(this.#currentRun);
+    this.#reader = { readdir, readFileAsText };
+    this.#runner = runner;
+    this.#index = {
+      realmURL: new URL(realm.url),
+      loader: Loader.createLoaderFromGlobal(),
+      ignoreMap: new URLMap(),
+      instances: new URLMap(),
+      modules: new Map(),
+      stats: {
+        instancesIndexed: 0,
+        instanceErrors: 0,
+        moduleErrors: 0,
+      },
+    };
   }
 
   get stats() {
-    return this.#currentRun.stats;
+    return this.#index.stats;
   }
 
   get loader() {
-    return this.#currentRun.loader;
+    return this.#index.loader;
+  }
+
+  get runState() {
+    return this.#index;
+  }
+
+  async run() {
+    await this.setupRunner(async () => {
+      if (!this.#fromScratch) {
+        throw new Error(`Index runner has not been registered`);
+      }
+      let current = await this.#fromScratch(this.#index.realmURL);
+      this.#index = {
+        instances: current.instances,
+        modules: current.modules,
+        ignoreMap: current.ignoreMap,
+        realmURL: current.realmURL,
+        stats: current.stats,
+        loader: Loader.createLoaderFromGlobal(),
+      };
+    });
   }
 
   async update(url: URL, opts?: { delete?: true }): Promise<void> {
-    this.#currentRun = await CurrentRun.incremental(
-      url,
-      opts?.delete ? "delete" : "update",
-      this.#currentRun
-    );
+    await this.setupRunner(async () => {
+      if (!this.#incremental) {
+        throw new Error(`Index runner has not been registered`);
+      }
+      let current = await this.#incremental(
+        this.#index,
+        url,
+        opts?.delete ? "delete" : "update"
+      );
+      this.#index = {
+        instances: current.instances,
```

Comment:
this clobbers the work that entry setter did--but also as part of invalidation, the index might have removed entries. so perhaps it is good to get the new copy here as it will not include any removed entries.... (unlike from scratch which will not remove any entries as part of indexing since it is starting "from scratch").